src/ImageSharp/Processing/Processors/Transforms/AffineProcessor.cs

```diff
@@ -0,0 +1,334 @@
+            Matrix3x2 matrix = this.GetProcessingMatrix(sourceBounds, this.targetRectangle);
+
+            if (this.Sampler is NearestNeighborResampler)
+            {
+                Parallel.For(
+                    0,
+                    height,
+                    configuration.ParallelOptions,
+                    y =>
+                    {
+                        Span<TPixel> destRow = destination.GetPixelRowSpan(y);
+
+                        for (int x = 0; x < width; x++)
+                        {
+                            var point = Point.Transform(new Point(x, y), matrix);
+                            if (sourceBounds.Contains(point.X, point.Y))
+                            {
+                                destRow[x] = source[point.X, point.Y];
+                            }
+                        }
+                    });
+
+                return;
+            }
+
+            int maxSourceX = source.Width - 1;
+            int maxSourceY = source.Height - 1;
+            (float radius, float scale, float ratio) xRadiusScale = this.GetSamplingRadius(source.Width, destination.Width);
+            (float radius, float scale, float ratio) yRadiusScale = this.GetSamplingRadius(source.Height, destination.Height);
+            float xScale = xRadiusScale.scale;
+            float yScale = yRadiusScale.scale;
+            var radius = new Vector2(xRadiusScale.radius, yRadiusScale.radius);
+            IResampler sampler = this.Sampler;
+            var maxSource = new Vector4(maxSourceX, maxSourceY, maxSourceX, maxSourceY);
+            int xLength = (int)MathF.Ceiling((radius.X * 2) + 2);
+            int yLength = (int)MathF.Ceiling((radius.Y * 2) + 2);
+
+            using (var yBuffer = new Buffer2D<float>(yLength, height))
+            using (var xBuffer = new Buffer2D<float>(xLength, height))
+            {
+                Parallel.For(
+                    0,
+                    height,
+                    configuration.ParallelOptions,
+                    y =>
+                        {
+                            Span<TPixel> destRow = destination.GetPixelRowSpan(y);
+                            Span<float> ySpan = yBuffer.GetRowSpan(y);
+                            Span<float> xSpan = xBuffer.GetRowSpan(y);
+
+                            for (int x = 0; x < width; x++)
+                            {
+                                // Use the single precision position to calculate correct bounding pixels
+                                // otherwise we get rogue pixels outside of the bounds.
+                                var point = Vector2.Transform(new Vector2(x, y), matrix);
+
+                                // Clamp sampling pixel radial extents to the source image edges
+                                Vector2 maxXY = point + radius;
+                                Vector2 minXY = point - radius;
+
+                                var extents = new Vector4(
```

Comment:
Is this correct in cases like `maxXY.X < 0`? Isn't it better to have:
```C#
var extents = new Vector4(
    MathF.Round(maxXY.X),
    MathF.Round(maxXY.Y),
    MathF.Round(minXY.X),
    MathF.Round(minXY.Y));
```

Reply:
No, the extants are used for getting the sub-pixel values to pass to the resampler. If you round them that's how you end up with the crappy output I had initially.